File: src/main/java/EiManualMutateGuidance.java
import edu.berkeley.cs.jqf.fuzz.ei.ExecutionIndex;
import edu.berkeley.cs.jqf.fuzz.guidance.Guidance;
import edu.berkeley.cs.jqf.fuzz.guidance.GuidanceException;
import edu.berkeley.cs.jqf.fuzz.guidance.Result;
import edu.berkeley.cs.jqf.instrument.tracing.SingleSnoop;
import edu.berkeley.cs.jqf.instrument.tracing.events.CallEvent;
import edu.berkeley.cs.jqf.instrument.tracing.events.ReturnEvent;
import edu.berkeley.cs.jqf.instrument.tracing.events.TraceEvent;

import java.io.InputStream;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.Random;
import java.util.function.Consumer;
import java.util.stream.Collectors;

/**
 * Analogous to a ExecutionIndexGuidance, but is backed by the above eiMap.
 * Needs to record coverage for EI to work.
 * Run only once.
 */
public class EiManualMutateGuidance implements Guidance {
    private Thread appThread = null; // Ensures only one thread
    private TraceEvent lastEvent = null;
    private EiState eiState = new EiState();

    public LinkedHashMap<ExecutionIndex, EiData> eiMap;
    private final Random rng;

    public EiManualMutateGuidance(LinkedHashMap<ExecutionIndex, EiData> eiMap, Random rng) {
        this.eiMap = eiMap;
        this.rng = rng;
    }

    public void reset() {
        eiState = new EiState();
    }

    /**
     * @return the full stack trace of the current EI state
     * should be invoked after getExecutionIndex was called on lastEvent
     */
    private String getFullStackTrace() {
        String[] lines = new String[eiState.depth + 1];
        // Manual copy because we don't care about counts
        for (int i = 0; i < eiState.depth + 1; i++) {
            int iid = eiState.rollingIndex[2 * i];
            int count = eiState.rollingIndex[2 * i + 1];
            lines[i] = "count " + count + " " + Server.eventStrings.get(iid);
        }
        return Arrays.stream(lines).collect(Collectors.joining(" || "));
    }

    @Override
    public InputStream getInput() throws IllegalStateException, GuidanceException {
        return new InputStream() {
            @Override
            public int read() {
                if (lastEvent == null) {
                    throw new GuidanceException("Could not compute execution index; no instrumentation?");
                }
                // Get the execution index of the last event
                ExecutionIndex executionIndex = eiState.getExecutionIndex(lastEvent);
                System.out.println("\tREAD " + Server.eventToString(lastEvent));
                // Attempt to get a value from the map, or else generate a random value
                return eiMap.computeIfAbsent(
                        executionIndex,
                        ei -> new EiData(getFullStackTrace(), rng.nextInt(256))
                ).choice;
            }
        };
    }

    @Override
    public boolean hasInput() {
        return true;
    }

    @Override
    public void handleResult(Result result, Throwable throwable) throws GuidanceException {
        System.out.println("\tHANDLE RESULT");
    }

    @Override
    public Consumer<TraceEvent> generateCallBack(Thread thread) {
        if (appThread != null) {
            throw new IllegalStateException("Guidance must run on main thread");
        }
        appThread = thread;
        String entryPoint = SingleSnoop.entryPoints.get(thread);
        if (entryPoint == null) {
            throw new IllegalStateException("Guidance must be able to determine entry point");
        }
        return this::handleEvent;
    }

    private boolean isTracking = false;
    private void handleEvent(TraceEvent e) {
        // Needed to cache stack traces
        String contents = Server.eventToString(e);
//            System.out.println("BEGIN VISIT");
        if (e instanceof CallEvent) {
            if (e.getContainingMethodName().equals("runGenerator")) {
//            if (((CallEvent) e).getInvokedMethodName().equals("Server#dummy()V")) {
                isTracking = true;
            }
            String trackedString = isTracking ? "*tracked" : "untracked";
            System.out.println("CALL " + trackedString + ": " + contents);
        } else if (e instanceof ReturnEvent) {
            String evString = e.getContainingClass() + "#" + e.getContainingMethodName();
//                if (evString.equals("com/pholser/junit/quickcheck/internal/GeometricDistribution#<init>")) {
//                if (evString.equals("Server#runGenerator")) {
            if (evString.contains("dummy")) {
                isTracking = false;
            }
            String trackedString = isTracking ? "*tracked" : "untracked";
            System.out.println("RET " + trackedString + ": " + evString);
        } else {
            String trackedString = isTracking ? "*tracked" : "untracked";
            System.out.println("OTHER " + trackedString + ": " + contents);
        }
        if (isTracking) {
            e.applyVisitor(eiState);
        }
        lastEvent = e;
//            System.out.println("END VISIT");
    }
}

File: src/main/kotlin/EiManualMutateGuidance.kt
import edu.berkeley.cs.jqf.fuzz.guidance.Guidance
import edu.berkeley.cs.jqf.fuzz.guidance.Result
import edu.berkeley.cs.jqf.instrument.tracing.events.TraceEvent
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.Closeable
import java.io.File
import java.io.InputStream
import java.util.*
import java.util.function.Consumer

enum class GuidanceMode {
    GENERATE_INPUT,
    REPRO_FILE,
    COLLECT_COV,
}

/**
 * A modified version of ExecutionIndexingGuidance that doesn't use ExecutionIndex. Go figure.
 *
 * ExecutionIndex is eschewed in favor of java.lang.Thread.getCurrentStackTrace, which is more reliable for our purposes
 * because we don't specifically need EI, and need to leave ASM out of -Xbootclasspath in order to allow Jacoco to
 * instrument and collect coverage on the fly.
 *
 * This class also doubles as a repro guidance and also a collector of test coverage, because there's some funky
 * thread stuff going on.
 */
class EiManualMutateGuidance<T>(rng: Random) : Guidance {
    var annotatingRandomSource: AnnotatingRandomSource? = null

    private var mode = GuidanceMode.GENERATE_INPUT

    private var hasRun = false

    val fuzzState = FuzzState(this, rng)
    val history: FuzzHistory<T> by fuzzState::history

    /**
     * During a repro run, the original choice map is saved here. Any values that did not appear in the choice run
     * are thereby preserved for diffing purposes.
     *
     * Outside a repro run, this should be null. Moving the original map to here instead of making this the repro map
     * and performing a null check hopefully saves cycles, although maybe the JIT would've saved them anyway.
     */
    private var reproBackupChoiceMap: ChoiceMap? = null

    /**
     * A stream of integers that will be consumed to fill values in the EI map.
     * Should be null when not performing repro runs.
     */
    var reproValues: Iterator<Int>? = null

    val isInReproMode: Boolean get() = mode == GuidanceMode.REPRO_FILE
    val isInCollectMode: Boolean get() = mode == GuidanceMode.COLLECT_COV

    /**
     * Sets the guidance to read input bytes from the provided file. This automatically calls `reset` to start
     * a new run, and clears the existing EI map.
     *
     * Until the returned closeable is closed, this guidance can only be used for repro; otherwise, the guidance
     * will read from saved values in the map.
     */
    fun reproWithFile(file: File): Closeable {
        this.reset()
        mode = GuidanceMode.REPRO_FILE
        reproBackupChoiceMap = fuzzState.choiceMap
        fuzzState.choiceMap = linkedMapOf()
        reproValues = file.readBytes().asSequence().map { it.toInt() and 0xFF }.iterator()
        return Closeable {
            // Careful with ordering here - we want the new choice map to override the values in the backed up map
            reproBackupChoiceMap!!.putAll(fuzzState.choiceMap)
            fuzzState.choiceMap = reproBackupChoiceMap!!
            reproBackupChoiceMap = null
            reproValues = null
            mode = GuidanceMode.GENERATE_INPUT
        }
    }

    /**
     * Sets the guidance to do nothing but collect test coverage; no inputs should ever be generated in this mode.
     */
    fun collectTestCov(): Closeable {
        mode = GuidanceMode.COLLECT_COV
        return Closeable {
            mode = GuidanceMode.GENERATE_INPUT
        }
    }

    fun reset() {
        hasRun = false
        fuzzState.resetForNewRun()
    }

    /**
     * When this object is not in repro mode (invoked through `reproWithFile`), this will read existing bytes from
     * the choice map and generate new bytes as necessary. In repro mode, this will read the next byte in the passed
     * in iterator.
     */
    override fun getInput(): InputStream {
        return object : InputStream() {
            override fun read(): Int {
                check(!isInCollectMode) { "Illegal attempt to request bytes while in coverage collection mode" }
                val info = annotatingRandomSource!!.consumeNextStackTraceInfo()
                log("\tREAD $info")
                return fuzzState.add(info)
            }
        }
    }

    override fun hasInput(): Boolean {
        return !hasRun
    }

    override fun handleResult(result: Result, throwable: Throwable) {
        hasRun = true
    }

    override fun generateCallBack(thread: Thread): Consumer<TraceEvent> {
        // Do nothing lol
        return Consumer {}
    }

    private fun log(msg: String) {
        if (verbose) {
            println(msg)
        }
    }
    /**
     * Produces the sequence of bytes produced by the most recent input.
     *
     * This emulates the behavior of Zest's ExecutionIndexingGuidance.MappedInput, which
     * simply produces all the bytes in the order that they were requested.
     */
    private fun getLastRunBytes(): ByteArray {
        return fuzzState.usedThisRun.map { k -> fuzzState.choiceMap[k]!!.toByte() }.toByteArray()
    }

    fun writeLastRun(dest: File) {
        dest.writeBytes(getLastRunBytes())
    }

    fun writeSessionHistory(dest: File) {
        val data = fuzzState.history
        dest.writeText(Json { prettyPrint = true }.encodeToString(data))
    }

    fun loadSessionHistory(src: File) {
        val newHistory: FuzzHistory<T> = Json.decodeFromString(src.readText())
        fuzzState.reloadFromHistory(newHistory)
    }

    companion object {
        private const val verbose = false
    }
}